python/ray/util/check_serialize.py

```diff
@@ -0,0 +1,205 @@
+"""A utility for debugging serialization issues."""
+
+import inspect
+import ray.cloudpickle as cp
+import colorama
+from contextlib import contextmanager
+
+
+@contextmanager
+def _indent(printer):
+    printer.level += 1
+    yield
+    printer.level -= 1
+
+
+class _Printer:
+    def __init__(self):
+        self.level = 0
+
+    def indent(self):
+        return _indent(self)
+
+    def print(self, msg):
+        indent = "    " * self.level
+        print(indent + msg)
+
+
+_printer = _Printer()
+
+
+class FailureTuple:
+    def __init__(self, obj, name, parent):
+        self.obj = obj
+        self.name = name
+        self.parent = parent
+
+    def __repr__(self):
+        return f"FailTuple({self.name} [obj={self.obj}, parent={self.parent}])"
+
+
+def _inspect_func_serialization(base_obj, depth, parent, failure_set):
+    assert inspect.isfunction(base_obj)
+    closure = inspect.getclosurevars(base_obj)
+    found = False
+    if closure.globals:
+        _printer.print(f"Detected {len(closure.globals)} global variables. "
+                       "Checking serializability...")
+
+        with _printer.indent():
+            for name, obj in closure.globals.items():
+                serializable, _ = inspect_serializability(
+                    obj,
+                    name=name,
+                    parent=parent,
+                    depth=depth - 1,
+                    _failure_set=failure_set)
+                found = found or not serializable
+                if found:
+                    break
+
+    if closure.nonlocals:
+        _printer.print(
+            f"Detected {len(closure.nonlocals)} nonlocal variables. "
+            "Checking serializability...")
+        with _printer.indent():
+            for name, obj in closure.nonlocals.items():
+                serializable, _ = inspect_serializability(
+                    obj,
+                    name=name,
+                    parent=parent,
+                    depth=depth - 1,
+                    _failure_set=failure_set)
+                found = found or not serializable
+                if found:
+                    break
+    if not found:
+        _printer.print(
+            f"WARNING: Did not find non-serializable object in {base_obj}. "
+            "This may be an oversight.")
+    return found
+
+
+def _inspect_generic_serialization(base_obj, depth, parent, failure_set):
+    assert not inspect.isfunction(base_obj)
+    functions = inspect.getmembers(base_obj, predicate=inspect.isfunction)
+    found = False
+    with _printer.indent():
+        for name, obj in functions:
+            serializable, _ = inspect_serializability(
+                obj,
+                name=name,
+                parent=parent,
+                depth=depth - 1,
+                _failure_set=failure_set)
+            found = found or not serializable
+            if found:
+                break
```

Comment:
Why do we break here? Shouldn't we try to identify all unserializable members?

Reply:
If not, should we only check other members (from line 100) if we didn't find anything before?

Reply:
We break because it's a bit difficult to print out something legible if we identify all (multiple objects can reference the same underlying culprit).

Reply:
It's a DFS approach

Reply:
Ok, thanks